dataset/src/main/scala/frameless/TypedEncoder.scala

```diff
@@ -196,7 +196,7 @@ object TypedEncoder extends LowPriorityTypedEncoder {
     def targetDataType: DataType = underlying.targetDataType
 
     def constructor(): Expression = {
-      WrapOption(underlying.constructor())
+      WrapOption(underlying.constructor(), underlying.sourceDataType)
```

Comment:
@kanterov Do you think that makes sense? (Tests do not pass with `underlying.targetDataType` instead)